### DISCUSSION TITLE
Integrating Hourly Energy Pricing into Powerwall-Dashboard

### DISCUSSION THREAD

**Discussion by SCHibbard** [jasonacox/Powerwall-Dashboard#466] [Show and tell]
@jasonacox suggested I post some dashboard panels I have added to Powerwall-Dashboard.  The first set is for energy pricing. Commonwealth Edison, the main Northern Illinois energy provider, offers hourly pricing rates (https://hourlypricing.comed.com). They also offer an API to pull current pricing (see https://hourlypricing.comed.com/hp-api). They publish pricing updates every 5-minutes.  The rate used for billing is based on an hourly average of the 5-minute prices (:00 to :55) for the hour the energy was consumed. ComEd’s API offers an Hourly Average, but the value is incorrect, as it calculates the past hour average, not top of hour to 55 past.  So just the 5-minute value is pulled from the API, an hourly average is calculated from :00 to :55 and applied to energy used/credited/saved for that same hour.  The Grafana Panels currently in use to display results are below. Note that are a bunch of other charges in a ComEd bill that are based mainly on KW consumption, crediting (if on net metering), or percentage of bill (taxes). Unfortunately many do not have a month-to-month rate consistency that can be reduced to a calculation.  I wish the bills were as cheap as the one panel shows, but they are normally 5-7x that (!) Plan to investigate that more one of these days.

![ComEd RT Pricing Panel Example](https://github.com/jasonacox/Powerwall-Dashboard/assets/40345296/683053b4-fa18-4cdc-add2-71e7490113ed)

![ComEd Costs-Credits-Savings Panel Example](https://github.com/jasonacox/Powerwall-Dashboard/assets/40345296/61c30cbd-7c88-451c-8afc-9dc2d031d50a)

![ComEd Monthly Costs+Credits Panel Example](https://github.com/jasonacox/Powerwall-Dashboard/assets/40345296/75bfdafc-e8e9-4e70-9014-a8ce6eea9f61)

**Comment by petersloan** [comment]
Hi @SCHibbard 

I am also in the Chicago area on the hourly pricing plan (since AMI meter installed in 2015) and just got my 2x PW3 + enphase panels w/IQ8X inverters installed awaiting final inspection.  I currently ingest the hourly pricing data into an influx DB for historical consumption/price monitoring but looking to get more sophisticated now that I will be generating credits.  I presume you setup a separate datasource? Did you end up leveraging the [Powerwall-Dashboard tools->mysql connector](https://github.com/jasonacox/Powerwall-Dashboard/tree/fbeda2ed0027dd13ae840d789d6941b21c5b487d/tools/mysql) or did you go a different route.  Anything you are willing to share (such as schema/grafana queries) or even potentially collaborate on would be appreciated.

Thanks!

**Reply by SCHibbard** [reply]
Hi Peter:  I ingest the data into the same Powerwall influx database created by the Powerwall Dashboard used for the rest of the Powerwall & weather data. I created a python script that runs as another Docker Container. It polls ComEd every 5 minutes and saves the current 5-minute price.  I don't save the hourly average they publish, due to the inaccuracies noted above. Instead, at the top of the hour it calculates the hourly average of the previous hour. That value is then applied to the various KwH values (Grid, PW, Solar) to generate the values used in the panels above.  I have on my list to clean up the code to make it publishable here, but I first wanted to tackle the problem of all the other charges ComEd tacks on, to try and get the monthly costs (last panel above) a little more accurate.  Been procrastinating that challenge!  They publish the rate in the bills, but many of them vary from month to month. The "Distribution Facility Charge", is a good portion of the bill.  Over a one year period, it went as low as 0.04065 to as high as 0.04777, That one alone contributes a lot to the bill total. I guess I could just use an average of the year since it is only about a 20% swing and be done with it, but I was hoping to find a way to pull it online.  I created an account at PJM (the data supplier to ComEd & others), but have not had time to do any investigation into what they can offer there. Any help in deciphering those Delivery, Taxes, and Fees charges would be welcome!!

**Comment by petersloan** [comment]
@SCHibbard 

For folks like myself with sufficient battery storage that will be operating in self powered mode, I would likely value my generation at the fixed rate price since I can use my generated power at any time.  So I would compute savings = (ComEd Generated Credit) + (kWh_locally_consumed * (Distribution_Facility_Charge + IL_Elec_Dist_Charge + PriceToCompareRate))  (See below for where I get these values).  I am hoping I can programmatically get the ComEd Generated Credit from their website but if not, its one value I would have to inject into InfluxDB monthly and the other values only change year (Starting in June I believe).  As for taxes and fees, I don't think the data entry is worth it and maybe a timeseries the latest fixed percentage can be used instead.  For example, with my bills, it appears that 25% of the computed "savings" would get you pretty close.  Anyways, on with the data dump.....

In my mind, essentially what you are trying to do is come up with two things (savings and actual cost) and will need to balance between manual data entry and automation.  

### Savings
This is the "value" that your solar system provided.  I think there are two ways to think about this.  If you are on the BHES (ComEd's TOU plan), you can actually use the value of the credit you generated that month since you are not credited kWH but rather the cash value at the time of generation.  Until my system is operational and starts to generate $ based credits, I won't know if this will be exposed anywhere in the various ComEd apis.  There is a site that the NREL referenced that has detailed rate information APIs but appears to be horribly out of date (https://apps.openei.org/services/doc/rest/util_rates/?version=7) after I signed up for key to look around.  Last update seems around 2016 and the source pages for the rates are all in the old site format (looks like you used to be able to get your rate information right from the site once you logged in).  

#### Electric Costs == Generation + Transmission
The other option or if you are on the standard BES plan is to lookup the Price To Compare price as this factors in both electricity cost and transmission cost.  [PlugInIllinois](https://plugin.illinois.gov/understanding-the-price-to-compare/price-to-compare-comed.html) lists the current Fixed Rate Electric cost + Transmission cost.

#### Delivery == ComEd's last mile distribution
Go to https://www.comed.com/current-rates-tariffs then click on your appropriate rate tariff and select A Guide to the Retail Customer’s Billed Delivery Service Charges (PDF)  And its a PDF printout of a excel doc because OF COURSE.... so you will need a process to manually key this in every 3 months.  The last column matches up with my bill.
Current rates can be found [here](https://azure-na-assets.contentstack.com/v3/assets/blt3ebb3fed6084be2a/blt7904befea93c3525/665e1aceb5316f000dc7347e/A_Guide_to_the_Retail_Customer_s_Billed_Delivery_Service_Charges.pdf?branch=prod_alias).  

#### Tariff and Fees
Each line item in this section has its own document.  If you go to https://www.comed.com/current-rates-tariffs you can see the PDFs corresponding to the actual adjustments although the docs can be named slightly differently so human involvement will be required.

### Actual costs
This is what ComED is actually charging you.

I was hoping to dig around find a way to get my bill in a format that is consumable by an API.  ComEd actually uses Oracles Opower platform which has the capability to do such things however it appears that each utility has their own instance which has been customized to their own needs and wants.  ComEd does not seem to have any endpoints that I can find that expose line item details however I am not sure that is exactly necessary.  There is an endpoint that exposes a massive JSON with just about everything ComEd knows about you and powers the My Bill Details of the ComEd Dashboard.  The tricky part is figuring out how to get to it.

I was happy to stumble on a python module called [Opower](https://github.com/tronikos/opower) which is used by Home Assistant.  I was hoping I could use it directly but as its a module designed to do one thing (get very basic consumption metrics), I couldn't easily leverage as a library import with my limited python skills so I opted to look how the comed (aka excelon) connector is implemented.

I apologize for the python that is about to follow as I am sure it will make more experienced folks eyes bleed. Being more of a C# developer, I never really spent much time in python other than quick and dirty things.  

Most of the steps required to actually get logged in and get all the appropriate bearer/csrf/etc tokens was from the [Exelon specific OPower connector](https://github.com/tronikos/opower/blob/8b2fa4550730ceda0e69f6ab544f2c476e0b34de/src/opower/utilities/exelon.py).  You just need to set the three variables at the top with your account specific credentials and account number ( I didn't feel like implementing account lookup for this quick POC ).  All the usual disclaimers around lack of robust error checking, handling multiple accounts, non residential accounts, MFA, or other edge cases I can't even think of.  Hopefully this could be a jumping off point to something more robust.  If there is a way to directly leverage the connector, that would be awesome as I am sure that project will be much better at keeping it up to date as the site changes (they implemented significant changes earlier this year that resulted in them being down for well over a week if I recall)

I would be really curious what this looks like for someone that has a month of production credits?  

```python
# USER EDIT VARS
accountid = "0000000000" #must be quoted, get this from your bill PDF or the My Account page of the dashboard
username = 'comedcust@domain.com' #What you use to login to comed's site
password = 'superdupersecrets'
# END USER EDIT VARS

import requests
import json
import re
import urllib3

USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
url = 'https://secure.comed.com/Pages/Login.aspx?/login'
login_domain = 'secure.comed.com'
headers = {'User-Agent': USER_AGENT}

session = requests.Session()
response = session.get(url, headers=headers)
result = response.text

parsedResponseUrl = urllib3.util.parse_url(response.url)

if parsedResponseUrl.path.endswith("/authorize"):
    settings_match = re.search(r"var SETTINGS = ({.*});", result)
    assert settings_match
    settings = json.loads(settings_match.group(1))
    login_post_domain = parsedResponseUrl.host
    assert login_post_domain
    
    response = session.post(
        "https://" + login_post_domain + settings["hosts"]["tenant"] + "/SelfAsserted",
        params={
            "tx": settings["transId"],
            "p": settings["hosts"]["policy"]
        },
        data={
            "request_type": "RESPONSE",
            "signInName": username,
            "password": password
        },
        headers={
            "X-CSRF-TOKEN": settings["csrf"],
            "User-Agent": USER_AGENT
        }
    )

    result_json = json.loads(response.text)

    if result_json["status"] != "200":
        print("Login Failed")
        print(result_json["message"])
    
    response = session.get(
        "https://"
        + login_post_domain
        + settings["hosts"]["tenant"]
        + "/api/"
        + settings["api"]
        + "/confirmed",
        params={
            "rememberMe": settings["config"]["enableRememberMe"],
            "csrf_token": settings["csrf"],
            "tx": settings["transId"],
            "p": settings["hosts"]["policy"],
            "diags": json.dumps(
                {
                    "pageViewId": settings["pageViewId"],
                    "pageId": settings["api"],
                    "trace": []
                }
            )
        },
        headers={"User-Agent": USER_AGENT}
    )

    # We are now logged in but have to acquire a bearer token
    response = session.get(
        "https://"
        + login_domain
        + "/api/Services/MyAccountService.svc/GetSession",
        headers={"User-Agent": USER_AGENT}
    )

    result_json = json.loads(response.text)
    bearer_token = result_json["token"]
    #print(json.dumps(result_json, indent=2))

    response = session.get(
        "https://" + login_domain 
        + "/.euapi/mobile/custom/auth/accounts/"
        + accountid
        + "?payments=false",
        headers={
            "User-Agent": USER_AGENT,
            "Authorization": "Bearer " + bearer_token
        }
    )

    accountJson = json.loads(response.text)
    #print(json.dumps(accountJson))
    #print(json.dumps(accountJson["data"]["BillingInfo"], indent=2))
    
    print("Supply: $" + str(accountJson["data"]["BillingInfo"]["supplyCharges"]))
    print("Delivery: $" + str(accountJson["data"]["BillingInfo"]["deliveryCharges"]))
    print("Taxes And Fees: $" + str(accountJson["data"]["BillingInfo"]["taxesAndFees"]))
```

**Reply by petersloan** [reply]
I will also mention the meat of the data is in the opower database which uses graphql to expose the account level data.  It would be somewhat meaty to try to fully explore this but the exelon connector referenced above has examples on how to get the bearer token from comed to hit the opower endpoints.  Might be something I mess around with more when I have time.   I would be loathe to try and construct my own queries and instead would rely on discovering an existing request that essentially could be replayed.

https://cec.opower.com/ei/edge/apis/dsm-graphql-v1/cws/graphql

https://docs.oracle.com/en/industries/energy-water/integration-hub/graphql-user-guide/Content/GraphQL/buidlingAndExploring.htm

**Reply by petersloan** [reply]
Just for the hell of it, I started messing around with creating a bill parser.  Super rough and crude but technically could automate downloading the bill PDF and running it through the parser with all the bits I started assembling.  Got to start heading to bed now so this is in no way completed, just some rough proof of concept type work.  

There appears to be an API to get a list of bills `https://secure.comed.com/.euapi/mobile/custom/auth/accounts/{accoundID}/billing/history` and then another you can call to download the bill `https://secure.comed.com/.euapi/mobile/custom/auth/accounts/{accountID}/billing/{Bill YYYY-MM-DD}/pdf`

```python
import pypdf
import json
import re

#Path to Bill to Parse, relative paths work
billPdfPath = "BillImage.pdf"

def extract_charges_and_fees(text):
    #Extracts charges, fees, total kWh, and per kWh cost from the given text.

    charges_data = []
    chargeTypeCategories = ["SUPPLY", "DELIVERY", "TAXES&FEES"]
    chargeCategory = "" 
    
    for line in text.splitlines():
        # We want to persist chargeCategory as we enumerate so we don't reset it between lines
        chargeDesc = ""
        kwhUsed = ""
        perKwhPrice = ""
        total = ""
        lineChunks = line.split();
        numOfLineChunks = len(lineChunks)
        
        if numOfLineChunks == 0:
            continue
        elif numOfLineChunks == 1:
            #skip as this is an info line
            continue
        elif numOfLineChunks == 2:
            # no kwh info
            for chargeTypeCategory in chargeTypeCategories:
                if line.startswith(chargeTypeCategory):
                    chargeCategory = chargeTypeCategory
                    break
            chargeDesc = lineChunks[0]
            total = lineChunks[1]
        elif numOfLineChunks == 3:
            # kwh info
            chargeDesc = lineChunks[0]
            total = lineChunks[2]
            if 'X' in lineChunks[1]:
                kwhInfoChunks = lineChunks[1].split('X')
                kwhUsed = kwhInfoChunks[0].lower().replace('kwh','')
                perKwhPrice = kwhInfoChunks[1]

        charges_data.append({"ChargeTypeCategory": chargeCategory, "ChargeDescription": chargeDesc, "KwhUsed": kwhUsed, "PerkWhPrice": perKwhPrice, "Total": total})

    return charges_data
        
            


def extract_section(text, start_marker, end_marker):
    start_index = text.find(start_marker)
    end_index = text.find(end_marker, start_index + len(start_marker))

    if start_index != -1 and end_index != -1:
        return text[start_index + len(start_marker):end_index]
    else:
        return None


with open(billPdfPath, 'rb') as pdf_file:
    pdf_reader = pypdf.PdfReader(pdf_file)
    num_pages = len(pdf_reader.pages)
    page = pdf_reader.pages[1]
    text = page.extract_text()
    
    billText = extract_section(text, "CHARGEDETAILS", "UPDATES")
    billObject = extract_charges_and_fees(billText)
    billJson = json.dumps(billObject, indent=2)
    print(billJson)


```

**Reply by SCHibbard** [reply]
@petersloan:

You dug up some good info!  A couple comments:

I assume by “I would likely value my generation at the fixed rate price since I can use my generated power at any time” you are suggesting you’d just use the summer or winter BES cents/kWh instead of the BESH hourly average. I am of the opinion that it is much more accurate to use the actual hourly rate.  My system has been up for over a year now.  I have 2 EVs, so I can drain the PWs if both my wife & I put on the miles during the day, but normally not.  In the summer cycle we are in now, the hourly price is normally high during the day, low at night.  So, when solar is producing, you’re saving quite a bit.  The batteries are riding through the overnight, but it is at a much lower BES cents/kWh rate.

Pulling the BESH rates from their open API at https://hourlypricing.comed.com/hp-api is fairly simple, just the aforementioned problem of the hourly rate not being what they really use for billing.  The power used values store in PWDB actually align very well with what ComEd says I’ve used at https://secure.comed.com/MyAccount/MyBillUsage/pages/secure/ViewMyUsage.aspx - normally within 1% of each other.  I chalk that off to efficiency losses in the PW Gateway, and/or jitter in times between ComEd and the Gateway.  So, grabbing the 5-minute value from the above API, calculating the average from :00 to :55, and applying it to the various kWh values already in PWDB (grid, solar, PW), gives pretty good results, albeit just raw Supply Charges, not transmission service charges, etc.  Here’s part of a comparison I did after the ComEd Docker container was up & running:

![image](https://github.com/user-attachments/assets/a55934aa-c085-4e38-8c96-6335c8c1ecbd)

ComEd reported price from above API tracks perfect with the calculated hourly average from the program.  Accuracy of what ComEd says I used or generated is within a percent or 2.  The issue is net cost per hour. What ComEd reports appears to include all the items under Supply & Delivery on a bill (just not Taxes & Fees).  The program only has Supply Charge so far.

I am aware of the Delivery & Tariff documentation, but like you said, it involves manual maintenance, and I have been hoping to avoid that.  I was not aware of the Opower info – good find!  I’ll dig into that and see what can be automated with it.  One immediately nice thing is it will provide bill date-range, which ComEd is not consistent at!

By the way, I added alerts in Grafana when pricing is going high – they track ComEd’s text messages, actually faster.  Also added one when it goes negative … plan is to tie that to charging the cars if they are home & plugged in & the PWs are full.

I’m going to be out of pocket the next couple of weeks at a conference, so it may be a while before I can dig into this.

    ---Scott---

**Reply by petersloan** [reply]
Sounds good @SCHibbard I have a few commitments coming up as well so will chew on your response on the next few days.  

Just to answer your question about the BES vs BESH rates as to how I would value savings.  I have been storing that data influx since 2021 and even longer in a mysql backend.  I don't have an EV so for me, moving to BESH was about reducing and shifting my consumption in order to generate savings (keeping my bill below $100 while others were pushing $300+), something that has required pretty significant lifestyle changes and automation that polls the hourly pricing API and controls a bunch of smart switches and thermostat based on real time rate thresholds.  By resuming "normal" life, I am effectively getting the "value" of being on a fixed rate plan again where I don't have to worry about TOU price fluctuations by relying on Self Powered Mode.  If electric supply was at .20/kwh, all my high draw devices would be automatically powered down, something that I expect I won't need to do with the storage.  Hopefully that explains why my approach to savings valuation is different.  I guess if powered spiked to over $3.50/kwh like it did back in Dec 2022 I could value any self consumption at that rate but in reality if I didn't have the ability to throttle down grid consumption, I wouldn't be on the BESH plan in the first place. 

![image](https://github.com/user-attachments/assets/903b0b05-ba9f-4bf4-a3fe-4e74f22e1427)

**Comment by petersloan** [comment]
Just as a note, I have "rediscovered" the ability to download actual historical hourly pricing rates from ComEd combined with hourly consumption and cost via the hourlypricing website.  It does require a signin and then clicking `My Graphs` at the top and then there is a little download icon just right of `Monthly Electricity Usage and Charges`.

The web request returns a response with the header Content-Disposition of `attachment; filename="173382798928-TQIASDGFTU29.csv"`.  I haven't figured out the login flow because its easier to just import the file into my influx once a year or so but technically you could use this programmatically to historically "correct" your calculated hourly average.   I plan to store it separately only for the matter of record.

https://hourlypricing.comed.com/rrtpbillcompare/Servlet?type=report-hourly-deliver&account={ENCODED_ACCOUNT_ID}&dateStart=12/05/2021&dateEnd=10/17/2024

@SCHibbard I do want to thank you for your enphase dashboard.  It made a couple of shading issues along the bottom portion of some of my panels readily apparent.

A note for those that might not know...  if your panel is in landscape mode and you have a shadow that spans the horizontal portion of the panel, even if a few inches, good chance that panel won't produce at all based on where the bypass diodes are placed.

**Reply by SCHibbard** [reply]
@petersloan I had not noticed that download option at the hourlypricing website, thanks for the tip.  Your message reminded me I was going to clean up that python script for ComEd & publish here ... guess it fell of my task list.  Sorry.

Yes, I have a pear tree in the backyard that has been trimmed a few times after seeing what it does to production from the panels on the garage!

**Reply by petersloan** [reply]
I found another unsecured endpoint that allows you to query BESH finalized hourly charges however it does seem to have a several day lag.  Thinking about creating a new measurement for finalizedHourlyPricing in influx to track this.  Unrelatedly, I just signed up for DTOD today but haven't been able to locate the official declaration of TOU charges for delivery.

https://secure.comed.com/api/Services/FTCService.svc/GetBESHReport?beginDate=01%2F01%2F2026&endDate=01%2F05%2F2026&pricingType=RealTime

**Reply by JonMurphy** [reply]
Yes, it seems to be always a two day lag.

There is also a link for day ahead pricing and it becomes available near midnight
https://secure.comed.com/api/Services/FTCService.svc/GetBESHReport?beginDate=01%2F19%2F2026&endDate=01%2F19%2F2026&pricingType=DayAhead

What is "DTOD"?

**Reply by petersloan** [reply]
DTOD is the new [Delivery Time of Day](https://www.comed.com/ways-to-save/for-your-home/manage-my-energy/delivery-time-of-day-pricing) (confusingly also called DTOU) which is a time of use tariff specifically for the Distribution Charge as of Jan 2026.  Currently everyone has a the flat rate of $0.059/kwh but DTOD now offers a Off Peak, Shoulder, and Peak pricing.  This appears especially good if you have legacy net metering with hourly pricing and would add a little cushion for those periods where we get negative supply rates.  Doing the math, it appears I would stand to save an additional $140/yr by earning more monetary credits for distribution during peak solar.

There is a pretty comprehensive [FAQ](https://www.comed.com/ways-to-save/for-your-home/manage-my-energy/delivery-time-of-day-pricing/todp-FAQs) page as well although like I said, I cannot find the source for the canonical rates.  Citizens Utility Board also doesn't seem to have any pertinent information on this despite being actively involved in its development and ratification.

I also believe Ameren is offering something similar as it was legislated by the state of IL although I cannot find anything on their site.

Unofficial numbers I have found for Distribution Charge (Flat rate is $0.059/kwh)

- Off Peak (9p - 6a): $0.03/kwh
- Morning Shoulder (6a - 1p): $0.04/kwh
- Peak (1p - 7p): $0.107/kwh
- Evening Shoulder (7p - 9p): $0.04/kwh

**Comment by JonMurphy** [comment]
I am in the ComEd area also and looking forward your python script!  I have two PW3, a GW3, and 20 solar panels installed Oct 2024.  And also I have ComEd hourly pricing (ugh!).
As a side note, did you get hit with the 800% increase for capacity charge?  Very painful for me!

**Reply by SCHibbard** [reply]
Jon:  That fell through the cracks over the winter, and I have a feeling I won't be getting my head back into it until the weather turns again, sorry!  But I will!  Peter had noted that you could download your usage direct from ComEd instead of getting it from PWDB (which would probably be closer to the actual ComEd bill than using the PW data) ... but it can only be accessed if MFA is not used for ComEd account login, which I use, and MFA cannot be turned off once activated. Darn!  So the project will not use that, as I have a feeling I'm not the only one using MFA.  Have not seen the increases yet, but got the nastygram from them last week warning it was coming.  On the other hand, bill was $1.96 last month, with 2 EVs!

**Reply by JonMurphy** [reply]
My MFA for ComEd is also enabled.

My March bill (billed in April) was $28 (72 kWh in / 154 kWh out)
My April bill (billed in May) was $15 (18 kWh in / 184 kWh out)
My May bill (billed in June with the new increase) was $42 (38 kWh in / 245 kWh out).  The capacity charge alone went up by **$25 PER MONTH**!  big ouch!
I have one EV.

Would you be willing to share your ComEd 5 minute average to Powerwall-Dashboard  influx database script?  I was about to do the same but with bash script.

Let me know when you need helping testing.  I am not a python person but I'd be happy to help test!

**Reply by SCHibbard** [reply]
My capacity charge has been at 0.91048 - have not seen the June increase yet.  But I think the capacity charge is based on not only peak supply costs (when you used the power), but also historical usage and location.  I'm in the city, I'd expect it to be worse here! 

Have not looked at the code in a long time, but I know it is a hot mess. It's been running about a year now with no problems. I recall I though first I could just grab the hourly rate they post, until I realized that value was for the previous 12 5-minute values, NOT the previous hour's 5 minute average, which is what they use to calculate costs. That made it messier.  Again, if I recall correctly, I added values to the influx data base with the hourly total KwH reported by the PWs, and then use that with the calculated hourly cost. Might take some time, but I'll look at it and see if I can clean it up to the point of posting.

**Comment by JonMurphy** [comment]
Here is a ComEd 5 minute average to Powerwall-Dashboard influx database script.  It is written in bash (since it is easy for me!) and speed really isn't an issue.   This watches for new ComEd Hourly Average data every five minutes and retries if nothing found.  Feel free to offer your 2¢ worth (or 1 kWh's worth).

Output example:
```
 $ influx -port 8086 -database 'energy_data1' -execute 'SELECT * FROM comed_hour_price'
name: comed_hour_price
time                price top_hr ts_local
----                ----- ------ --------
1761427500000000000 2.3   false  2025-10-25 16:25:00 -0500
1761427800000000000 2.3   false  2025-10-25 16:30:00 -0500
1761428100000000000 2.2   false  2025-10-25 16:35:00 -0500
1761428400000000000 1.8   false  2025-10-25 16:40:00 -0500
1761428700000000000 1.8   false  2025-10-25 16:45:00 -0500
1761429000000000000 1.8   false  2025-10-25 16:50:00 -0500
```

Script:
```bash
#!/bin/bash
#===============================================================================
# Script Name   : comed_hourly_influx.sh
# Description   : Polls ComEd's Hourly Pricing API, parses the current
#                 average hourly price, and writes results to InfluxDB
#                 in line protocol format. Includes retry logic, age checks, and
#                 simple timestamp validation to ensure accurate hourly updates.
# Author        : Jon Murphy (with big AI assist)
# Created       : 2025-10-25
#
#===============================================================================
# Usage         : ./comed_hourly_influx.sh
# Notes         :
#   - Requires valid INFLUX_URL, INFLUX_DB, and API_URL settings.
#   - Designed for headless operation (e.g., Raspberry Pi).
#   - Implements safe integer math for time comparisons.
#===============================================================================
set -uo pipefail

# URL for ComEd Current Hour Average API & Day Ahead
API_URL="https://hourlypricing.comed.com/api?type=currenthouraverage"

# InfluxDB connection details
INFLUX_URL="http://192.168.65.225:8086"
INFLUX_DB="energy_data1"
INFLUX_MEASUREMENT="comed_hour_price"

# for scraping ComEd Day-Ahead Hourly Prices
LAST_SCRAPE_HASH=""
LAST_SCRAPE_RUN=""

# Timezone for formatting timestamps locally
TZ="America/Chicago"

# Tag name for logging context
tagName="comed_hourly1"

last_nanos=0		# Track the last timestamp successfully written to InfluxDB


########################### Functions ###############################

# Fetch current hour average and format as InfluxDB line protocol
get_hourly_average() {
    local api_out millisUTC price ts_sec min top_of_hour local_time
	local line_protocol

    # Fetch API data, capture errors in api_out
    api_out=$(curl --silent --show-error --fail "${API_URL}" 2>&1)

    # Return error if curl fails or response is empty
    if [[ $? -ne 0 || -z "${api_out}" ]]; then
        msg_log ERROR "curl fail getting data: ${api_out//$'\n'/}"
        return 1
    fi

    # Extract 'millisUTC' (timestamp) and 'price' from the JSON response using jq
    millisUTC=$(echo "${api_out}" | jq -r '.[0].millisUTC' 2>/dev/null)
    price=$(echo "${api_out}" | jq -r '.[0].price' 2>/dev/null)

    # Validate values are present and not null
    if [[ -z "${millisUTC}" || "${millisUTC}" == "null" || -z "${price}" || "${price}" == "null" ]]
    then
        msg_log ERROR "missing millisUTC or price in API response: ${api_out//$'\n'/}"
        return 1
    fi

	ts_sec=$(( millisUTC / 1000 ))	# Convert millisUTC to seconds for epoch

    # Extract minute component in UTC time to determine if data is for top of hour
    min=$(date -u -d "@${ts_sec}" +%M)
	[[ "${min}" == "00" ]] && top_of_hour=true || top_of_hour=false

    # Format a human-readable local time string for reference
    local_time=$(TZ="${TZ}" date -d "@${ts_sec}" '+%Y-%m-%d %H:%M:%S %z')

    # Construct InfluxDB line protocol string:
    #   measurement, 'top of hour', 'price', 'local time', and timestamp (ns)
	line_protocol="${INFLUX_MEASUREMENT} top_hr=${top_of_hour},"
	line_protocol+="price=${price},ts_local=\"${local_time}\" $((ts_sec * 1000000000))"

    # Output constructed line protocol for InfluxDB
    echo "${line_protocol}"
    return 0
}

msg_log () {
	logger --tag "${tagName}" "$*"
	if tty --silent ; then
		echo "$(date "+%b %d %H:%M:%S") ${tagName}:" "$*"
	fi
}

check_create_database() {
	local influx_url="$1"
	local influx_db="$2"
	local db_exists

	db_exists=$(curl --silent --get "${influx_url}/query" \
		--data-urlencode "q=SHOW DATABASES" | grep --word-regexp "\"${influx_db}\"")

	if [[ -z "${db_exists}" ]]; then
		msg_log WARN "database \"${influx_db}\" does not exist. Creating..."

		response=$(curl --silent "${influx_url}/query" --data-urlencode "q=CREATE DATABASE ${influx_db}")
		if echo "$response" | jq -e '.results[0].error' > /dev/null; then
			msg_log ERROR "failed to create database \"${influx_db}\""
			return 1
		else
			msg_log INFO "Database \"${influx_db}\" created"
			return 0
		fi
	fi

	return 0
}

write_to_influx() {
	local influx_url="$1"
	local influx_db="$2"
	local line_protocol="$3"

	# Perform the curl POST, capture response body + HTTP status code
	local response

	## maybe add --fail to curl?
	response=$(curl --silent \
		--show-error \
		--write-out "%{http_code}" \
		--request POST "${influx_url}/write?db=${influx_db}" \
		--data-binary "${line_protocol}")

	# Extract HTTP status code (last 3 chars)
	local http_code="${response: -3}"
	# Extract response body (everything except last 3 chars)
	local body="${response:0:-3}"

	if [[ "${http_code}" != "204" ]]; then
		msg_log ERROR "write failed with HTTP status ${http_code}"
		[[ -n "${body}" ]] && msg_log ERROR "response body: ${body}"
		return 1
	fi

	return 0
}

#	calculated seconds of sleep
sleep_seconds() {
	local interval_seconds=300		# Interval between runs = 5 minutes * 60 seconds = 300
	local offset_seconds=0			# 0 minute * 60 seconds for no offset

	local now=$( date +%s )
	local next=$(( ((now - offset_seconds) / interval_seconds + 1) * interval_seconds + offset_seconds ))
	echo $(( next - now ))
}

##############################     Main     ##############################

# Check if database exists
check_create_database "${INFLUX_URL}" "${INFLUX_DB}"
[[ $? -ne 0 ]] && msg_log ERROR "Database check/create failed." && exit 1

# Main infinite loop: sleep until next scheduled poll, then retry to get data
while true; do

    # Retry polling loop, up to N retries spaced by 30 seconds
    for attempt in {1..100}; do
        line_protocol=$(get_hourly_average)
        status=$?

		if [[ ${status} -eq 0 && -n ${line_protocol} ]]; then
			# Diagnostic output with timestamp, attempt count, and data
			msg_log DEBUG "attempt #${attempt} | line_protocol: ${line_protocol}"

			# Extract last field (epoch nanos timestamp) from line protocol string
			new_nanos=$(echo "${line_protocol}" | awk '{print $NF}')

			# Write only if timestamp advances and data is sufficiently fresh
			if (( new_nanos > last_nanos )); then
				# Write data point
				write_to_influx "${INFLUX_URL}" "${INFLUX_DB}" "${line_protocol}"
				if [[ $? -ne 0 ]] ; then
					msg_log ERROR "unable to write to InfluxDB"
					continue	# skip remain of "for loop" and go to next attempt
				fi
				msg_log INFO "InfluxDB Write successful"

				last_nanos="${new_nanos}" 		# Update last written timestamp
				break							# exit "for loop"
			else
				data_age_sec=$(( $(date +%s) - (new_nanos / 1000000000) ))
				#  Log that old or duplicate data was found; continue retrying
				msg_log WARN "Found old data (age of ${data_age_sec} sec) at attempt #${attempt}. Retry in 30 sec"
			fi
		fi
		sleep 30  # Delay before next retry attempt (30 seconds)
	done	#  for loop

	sleep_sec=$(sleep_seconds)		#  calculate sleep until next 5-minute mark
	msg_log INFO "Sleep for ${sleep_sec} sec"

    sleep "${sleep_sec}"					# Sleep until next poll moment

done		#  while loop
```

**Reply by SCHibbard** [reply]
Looks good!  I guess the weather is changing, and I will soon loose my excuse to procrastinate cleaning up the python container I use so it is suitable for publication.

**Comment by JonMurphy** [comment]
Concerning this in the OP:  "_The rate used for billing is based on an hourly average of the 5-minute prices (:00 to :55)_".  

The rate charge to ComEd HP users is the top of the hour rate, and none of the other ¢ published.  Any hourly average data from :05, :10 to :50, :55 is just an **guesstimate** of the final cost.  What is published at the top of the hour is the price for the **previous** hour.

**as a side:**  can you image purchasing gasoline like this?!?  "Sorry sir/madam, your final price and invoice will be printed at 01:07:30 PM when we calculate gasoline costs for the previous hour".

**Reply by SCHibbard** [reply]
Yup, and even more deceiving is what they publish in their API for hourly.  When I investigated it, the hourly average is based on the last 12 5-minute values ... NOT top of hour.  so their hourly value is of no use.

**Reply by JonMurphy** [reply]
> NOT top of hour. so their hourly value is of no use.

The only time they are of use is at the top of the hour.  Everything else is an estimate.

```
$ influx -database 'energy_data' -execute 'SELECT * FROM comed_hour_price WHERE top_hr=true'
name: comed_hour_price
time                price top_hr ts_local
----                ----- ------ --------
1761105600000000000 1.9   true   2025-10-21 23:00:00 -0500
1761109200000000000 1.9   true   2025-10-22 00:00:00 -0500
1761112800000000000 2.2   true   2025-10-22 01:00:00 -0500
1761116400000000000 1.6   true   2025-10-22 02:00:00 -0500
1761120000000000000 1.3   true   2025-10-22 03:00:00 -0500
1761123600000000000 1.3   true   2025-10-22 04:00:00 -0500
```
